### PR TITLE
Append extra plugin data in server list ping

### PR DESCRIPTION
**Pull request by TheMolkaPL**
PGM can now append extra information about itself in [server list ping](https://wiki.vg/Server_List_Ping). ~~This feature is toggleable in the configuration - switch `server-ping-data.enabled` to `false`.~~

SportPaper implementation: https://github.com/Electroid/SportPaper/pull/15
Related: https://github.com/ProjectEnyo/projectenyo.github.io/pull/4#issuecomment-569471250

Complete server ping response from my local server:

```json
{
    "description": "§7Â» §bAirship Battle §7Â«",
    "players": {
        "max": 20,
        "online": 0
    },
    "version": {
        "name": "SportPaper 1.8.8",
        "protocol": 47
    },
    "bukkit_extra": {
        "pgm": {
            "0": {
                "state": "idle",
                "duration": 0,
                "observers": 0,
                "participants": 0,
                "max_players": 100,
                "map": {
                    "slug": "airship_battle",
                    "name": "Airship Battle",
                    "version": "1.0.4",
                    "objective": "Leak lava from the enemy\u0027s obsidian core into the void.",
                    "authors": [
                        {
                            "uuid": "30e27366-0b14-4076-8f55-0819ece49ce3",
                            "name": "dewtroid"
                        }
                    ]
                },
                "next_map": {
                    "slug": "harb",
                    "name": "Harb",
                    "version": "1.3.5",
                    "objective": "Be the team with the least amount of deaths after 10 minutes.",
                    "authors": [
                        {
                            "uuid": "1379cb6e-f291-4498-9807-e636f9674ac0",
                            "name": "SH4D0W_HAWK"
                        },
                        {
                            "uuid": "ef4ea031-998f-4ec9-b7b6-1bdd428bcef8",
                            "name": "Plastix"
                        }
                    ]
                }
            }
        }
    }
}

```

### CODE REVIEW

**Comment by TheMolkaPL** [comment]
It failed to build because it was depending on SportPaper prior to https://github.com/Electroid/SportPaper/commit/d85121f17590656a46750a405e824a2824b3e122

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/runner/work/PGM/PGM/src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java:[39,28] cannot find symbol
  symbol:   method getOrCreateExtra(tc.oc.pgm.api.PGM)
  location: variable event of type org.bukkit.event.server.ServerListPingEvent
```

File: src/main/resources/config.yml [hunk 1]
```diff
@@ -92,6 +92,9 @@ motd:
     running: green                                       # Color designating a match which is currently running
     finished: red                                        # Color designating a match which has ended
 
+server-ping-data:
+  enabled: true
+
```
Comment (Electroid) [inline src/main/resources/config.yml]: TBH, we don't need a toggle for this. There's really no reason to disable it. Let's just always do it.

File: src/main/java/tc/oc/pgm/api/match/MatchPhase.java [hunk 2]
```diff
@@ -50,4 +51,9 @@ public boolean canTransitionTo(MatchPhase next) {
         throw new IllegalStateException("Unknown transition state for " + next);
     }
   }
+
+  @Override
+  public String toString() {
+    return this.name().toLowerCase(Locale.ROOT);
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/api/match/MatchPhase.java]: This seems like a breaking change? Why the change?

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  private final MatchManager matchManager;
+  private boolean ready;
+
+  public ServerPingDataListener(MatchManager matchManager) {
+    this.matchManager = checkNotNull(matchManager);
+  }
+
+  @EventHandler
+  public void onMatchLoad(MatchLoadEvent event) {
+    ready = true;
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: Use an `AtomicBoolean`

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchManager;
+import tc.oc.pgm.api.match.event.MatchLoadEvent;
+import tc.oc.pgm.map.Contributor;
+import tc.oc.pgm.map.MapInfo;
+import tc.oc.pgm.map.PGMMap;
+import tc.oc.pgm.rotation.PGMMapOrder;
+
+public class ServerPingDataListener implements Listener {
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: Isn't there already some `PingListener`? Can we have the Motd and this data all in one listener?

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  @EventHandler
+  public void onServerListPing(ServerListPingEvent event) {
+    if (!Config.ServerPingData.enabled() || !ready) return;
+
+    JsonObject root = event.getOrCreateExtra(PGM.get());
+    for (Match match : this.matchManager.getMatches()) {
+      JsonObject matchObject = new JsonObject();
+      this.serializeMatch(match, matchObject);
+      root.add(match.getId(), matchObject);
+    }
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: I think we should be caching this. We want server pings to be super fast.

Use a `Cache`

File: src/main/java/tc/oc/pgm/match/MatchManagerImpl.java [hunk 6]
```diff
@@ -45,7 +46,7 @@
   private final MapLibrary library;
   private final MapLoader loader;
 
-  private final Map<String, Match> matchById = new HashMap<>();
+  private final Map<String, Match> matchById = new ConcurrentHashMap<>();
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/match/MatchManagerImpl.java]: I presume this is because the `PingListener` is getting this on an async thread?

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java (hunk 4 above)
Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: There is `MotdListener`. https://github.com/TheMolkaPL/PGM-1/blob/de337077ef99a96c0fc4051bc6359061ad049d73/src/main/java/tc/oc/pgm/listeners/MotdListener.java

I felt that it is cleaner to keep them separated. https://github.com/TheMolkaPL/PGM-1/blob/de337077ef99a96c0fc4051bc6359061ad049d73/src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java

File: src/main/java/tc/oc/pgm/match/MatchManagerImpl.java (hunk 6 above)
Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/match/MatchManagerImpl.java]: exactly

File: src/main/java/tc/oc/pgm/api/match/MatchPhase.java (hunk 2 above)
Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/api/match/MatchPhase.java]: I checked all usages of `MatchPhase` and this is not breaking. I use this [here](https://github.com/TheMolkaPL/PGM-1/blob/de337077ef99a96c0fc4051bc6359061ad049d73/src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java#L53) to serialize `MatchPhase` (`starting` instead of `STARTING`, etc). I could do this in `ServerPingDataListener`, but I felt that doing this here is more universal.

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java (hunk 5 above)
Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: done

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  @EventHandler
+  public void onServerListPing(ServerListPingEvent event) {
+    if (!ready.get()) return;
+
+    JsonObject root = event.getOrCreateExtra(PGM.get());
+    for (Match match : this.matchManager.getMatches()) {
+      try {
+        root.add(match.getId(), this.matchCache.get(match));
+      } catch (ExecutionException e) {
+        throw new RuntimeException(e);
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: Log this and don't throw the error.

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  private final MatchManager matchManager;
+  private final AtomicBoolean ready;
+  private final LoadingCache<Match, JsonObject> matchCache;
+
+  public ServerPingDataListener(MatchManager matchManager) {
+    this.matchManager = checkNotNull(matchManager);
+    this.ready = new AtomicBoolean();
+    this.matchCache =
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(1L, TimeUnit.SECONDS)
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: Too aggressive, let's cache for 5 seconds.

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+import tc.oc.pgm.map.Contributor;
+import tc.oc.pgm.map.MapInfo;
+import tc.oc.pgm.map.PGMMap;
+import tc.oc.pgm.rotation.PGMMapOrder;
+
+public class ServerPingDataListener implements Listener {
+
+  private final MatchManager matchManager;
+  private final AtomicBoolean ready;
+  private final LoadingCache<Match, JsonObject> matchCache;
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: I'd rather not store a strong reference to the match, how about a string, which is the match id.

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+public class ServerPingDataListener implements Listener {
+
+  private final MatchManager matchManager;
+  private final Logger logger;
+  private final AtomicBoolean ready;
+  private final LoadingCache<Match, JsonObject> matchCache;
+
+  public ServerPingDataListener(MatchManager matchManager, Logger parentLogger) {
+    this.matchManager = checkNotNull(matchManager);
+    this.logger = ClassLogger.get(parentLogger, ServerPingDataListener.class);
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: checkNotNull(parentLogger)

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+  public void onServerListPing(ServerListPingEvent event) {
+    if (!ready.get()) return;
+
+    JsonObject root = event.getOrCreateExtra(PGM.get());
+    for (Match match : this.matchManager.getMatches()) {
+      String matchId = match.getId();
+      try {
+        root.add(matchId, this.matchCache.get(match));
+      } catch (ExecutionException e) {
+        this.logger.log(Level.SEVERE, "Could not cache match " + matchId, e);
```
Comment (Electroid) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: `Could not load server ping data for match: `

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java (hunk 10 above)
Comment (TheMolkaPL) [reply src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: It's nullable in `ClassLogger`.

https://github.com/TheMolkaPL/PGM-1/blob/b3b5a18b7be80789e3a6b8187606a9de9dc91eff/src/main/java/tc/oc/util/logging/ClassLogger.java#L62-L70

**Comment by Electroid** [comment]
Don't want to merge this yet, since if people don't update SportPaper things will break. Thinking about what's best to do here.

**Comment by TheMolkaPL** [comment]
Merged changes from master 👌

File: src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  @EventHandler
+  public void onMatchLoad(MatchLoadEvent event) {
+    ready.compareAndSet(false, true);
+  }
+
+  @EventHandler
+  public void onServerListPing(ServerListPingEvent event) {
+    if (!ready.get()) return;
+
+    JsonObject root = event.getOrCreateExtra(PGM.get());
```
Comment (Pablete1234) [inline src/main/java/tc/oc/pgm/listeners/ServerPingDataListener.java]: Can't you just try/catch this for the method not found exception/throwable, and return if it throws? AFAIK it's the only thing that uses a new sportbukkit method
